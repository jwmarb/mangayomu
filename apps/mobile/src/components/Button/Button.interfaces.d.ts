import { ButtonColors } from '@mangayomu/theme';
import React from 'react';
import { PressableProps, StyleProp, ViewStyle } from 'react-native';

export interface ButtonProps
  extends React.PropsWithChildren,
    Omit<PressableProps, 'style'> {
  label?: string;
  variant?: 'contained' | 'outline' | 'text';
  color?: ButtonColors;
  disabled?: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  icon?: React.ReactElement<any>;
  iconPlacement?: 'left' | 'right';
  sharp?: boolean;
  style?: StyleProp<ViewStyle>;
}
