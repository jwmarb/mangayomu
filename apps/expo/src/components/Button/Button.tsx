import { ButtonBase } from '@components/Button/ButtonBase/ButtonBase';
import { ButtonProps, ButtonTextProps } from '@components/Button/Button.interfaces';
import React from 'react';
import { ButtonContainer, ButtonText } from '@components/Button/Button.base';
import { Typography } from '@components/Typography';
import { ButtonBaseProps } from '@components/Button/ButtonBase/ButtonBase.interfaces';
import Spacer from '@components/Spacer';
import { useTheme } from 'styled-components/native';
import { Color } from '@theme/core';

const Button: React.FC<ButtonProps> = (props) => {
  const {
    onPress = () => void 0,
    onLongPress = () => void 0,
    color = 'primary',
    variant = 'text',
    title,
    disabled = false,
    expand = false,
    icon,
    iconPlacement = 'left',
    ...rest
  } = props;
  const passedProps: Required<Omit<ButtonBaseProps, 'opacity'>> = {
    color,
    variant,
    disabled,
    expand,
    onPress,
    onLongPress,
    round: false,
    square: false,
    useGestureHandler: false,
  };
  const buttonTextProps: ButtonTextProps = {
    color,
    variant,
    disabled,
    expand,
    round: false,
    square: false,
    useGestureHandler: false,
  };

  const theme = useTheme();

  const iconColor = React.useMemo(() => {
    if (disabled) return 'disabled';
    if (color instanceof Color) return Color.rgba(color.getContrastText().get(), color.getContrastText().get());
    if (variant === 'contained')
      return Color.rgba(
        theme.palette[color].main.getContrastText().get(),
        theme.palette[color].main.getContrastText().get()
      );
    return color;
  }, [color, disabled, variant]);

  return (
    <ButtonBase {...passedProps} {...rest}>
      <ButtonContainer expand={expand}>
        {icon ? (
          iconPlacement === 'left' ? (
            <>
              {React.cloneElement(icon, { size: 'small', color: iconColor })}
              {title.length > 0 && (
                <>
                  <Spacer x={1} />
                  <ButtonText {...buttonTextProps}>{title}</ButtonText>
                </>
              )}
            </>
          ) : (
            <>
              {title.length > 0 && (
                <>
                  <ButtonText {...buttonTextProps}>{title}</ButtonText>
                  <Spacer x={1} />
                </>
              )}
              {React.cloneElement(icon, { size: 'small', color: iconColor })}
            </>
          )
        ) : (
          <ButtonText {...buttonTextProps}>{title}</ButtonText>
        )}
      </ButtonContainer>
    </ButtonBase>
  );
};

export default Button;
