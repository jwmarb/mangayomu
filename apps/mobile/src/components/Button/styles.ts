import { ViewStyle } from 'react-native';
import type { ButtonVariants, ButtonColors } from '@/components/types';
import { createStyles } from '@/utils/theme';

export const variants = (color: ButtonColors) =>
  createStyles<Record<ButtonVariants, ViewStyle>>((theme) => ({
    contained: {
      backgroundColor: theme.palette[color].main,
    },
    outlined: {
      borderColor: theme.palette[color].main,
    },
    text: {},
  }));

export const styles = (variant: ButtonVariants, color: ButtonColors) =>
  createStyles((theme) => ({
    button: {
      paddingHorizontal: theme.style.size.xxl,
      paddingVertical: theme.style.size.l,
    },
    container: {
      borderWidth: 1.5,
      borderRadius: theme.style.borderRadius,
      borderColor:
        variant === 'outlined' ? theme.palette[color].main : 'transparent',
      overflow: 'hidden',
    },
  }));
