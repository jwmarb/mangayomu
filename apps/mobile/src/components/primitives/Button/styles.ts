import { ViewStyle } from 'react-native';
import type {
  ButtonVariants,
  ButtonColors,
} from '@/components/primitives/types';
import { createStyles } from '@/utils/theme';

export const variants = (color: ButtonColors, state: 'enabled' | 'disabled') =>
  createStyles<Record<ButtonVariants, ViewStyle>>((theme) => ({
    contained: {
      backgroundColor:
        state === 'disabled'
          ? theme.palette.action.disabled
          : theme.palette[color].main,
    },
    outlined: {
      borderColor:
        state === 'disabled'
          ? theme.palette.action.disabled
          : theme.palette[color].main,
    },
    text: {},
  }));

export const styles = (variant: ButtonVariants, color: ButtonColors) =>
  createStyles((theme) => ({
    button: {
      paddingHorizontal: theme.style.size.xxl,
      paddingVertical: theme.style.size.l,
    },
    container: {
      borderWidth: 1.5,
      borderRadius: theme.style.borderRadius,
      borderColor:
        variant === 'outlined' ? theme.palette[color].main : 'transparent',
      overflow: 'hidden',
    },
    pressable: {
      gap: theme.style.size.m,
      alignItems: 'center',
      flexDirection: 'row',
    },
    text: {
      flexGrow: 1,
    },
  }));
