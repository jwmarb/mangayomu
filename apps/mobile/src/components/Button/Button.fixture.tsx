import { ScrollView } from 'react-native';
import { useSelect } from 'react-cosmos/client';
import { SafeAreaView } from 'react-native-safe-area-context';
import Button from '@/components/Button';
import { createStyles } from '@/utils/theme';
import useStyles from '@/hooks/useStyles';
import useModeSelect from '@/hooks/useModeSelect';
import Contrast from '@/components/Contrast';
import { BUTTON_COLORS, ButtonColors } from '@/components/types';

const styles = createStyles((theme) => ({
  container: {
    flex: 1,
    backgroundColor: theme.palette.background.default,
  },
  content: {
    gap: theme.style.size.l,
    padding: theme.style.size.xl,
    justifyContent: 'center',
  },
  padding: {
    padding: theme.style.size.l,
  },
}));

export default function ButtonFixture() {
  const mode = useModeSelect();
  const style = useStyles(styles, mode);
  const [color] = useSelect('color', {
    defaultValue: 'primary',
    options: BUTTON_COLORS as unknown as string[],
  }) as unknown as [ButtonColors];
  const [textAlignment] = useSelect('textAlignment', {
    defaultValue: 'left',
    options: ['left', 'center', 'right'],
  });
  return (
    <SafeAreaView style={style.container}>
      <ScrollView contentContainerStyle={style.content}>
        <Contrast contrast={mode}>
          <Button
            title="Contained"
            textAlignment={textAlignment}
            variant="contained"
            color={color}
          />
          <Button
            title="Outlined"
            textAlignment={textAlignment}
            variant="outlined"
            color={color}
          />
          <Button title="Text" textAlignment={textAlignment} color={color} />
        </Contrast>
      </ScrollView>
    </SafeAreaView>
  );
}
