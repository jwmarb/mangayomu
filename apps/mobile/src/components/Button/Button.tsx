import React from 'react';
import { ButtonProps } from './Button.interfaces';
import { BaseButton } from 'react-native-gesture-handler';
import { moderateScale } from 'react-native-size-matters';
import Text from '@components/Text';
import { useTheme } from '@emotion/react';
import Stack from '@components/Stack';
import Box from '@components/Box/Box';

const Button: React.FC<ButtonProps> = (props) => {
  const {
    label,
    variant,
    color = 'primary',
    disabled = false,
    icon,
    iconPlacement = 'left',
    style,
    ...rest
  } = props;
  const theme = useTheme();
  return (
    <Box
      overflow="hidden"
      style={[
        style,
        {
          borderRadius: theme.style.borderRadius,
          backgroundColor:
            variant === 'contained'
              ? disabled
                ? theme.palette.text.disabled
                : theme.palette[color].main
              : undefined,
        },
      ]}
    >
      <BaseButton
        shouldCancelWhenOutside
        disallowInterruption
        enabled={!disabled}
        {...rest}
        rippleColor={theme.palette[color].ripple}
      >
        <Stack
          space={moderateScale(4)}
          py={moderateScale(12)}
          px={moderateScale(16)}
          flex-direction="row"
          align-items="center"
          justify-content="center"
          border-width="@theme"
          border-radius="@theme"
          style={{
            borderColor:
              variant === 'outline'
                ? disabled
                  ? theme.palette.text.disabled
                  : theme.helpers.getColor(color)
                : 'transparent',
          }}
        >
          {iconPlacement === 'left' &&
            icon &&
            React.cloneElement(icon, {
              color: disabled
                ? 'disabled'
                : variant === 'contained'
                ? {
                    custom: theme.helpers.getContrastText(
                      theme.palette[color].main,
                    ),
                  }
                : color,
            })}
          {label && (
            <Text
              variant="button"
              align="center"
              bold
              color={
                disabled
                  ? 'disabled'
                  : variant === 'contained'
                  ? {
                      custom: theme.helpers.getContrastText(
                        theme.palette[color].main,
                      ),
                    }
                  : color
              }
            >
              {label}
            </Text>
          )}
          {iconPlacement === 'right' &&
            icon &&
            React.cloneElement(icon, {
              color: disabled
                ? 'disabled'
                : variant === 'contained'
                ? {
                    custom: theme.helpers.getContrastText(
                      theme.palette[color].main,
                    ),
                  }
                : color,
            })}
        </Stack>
      </BaseButton>
    </Box>
  );
};

export default Button;
