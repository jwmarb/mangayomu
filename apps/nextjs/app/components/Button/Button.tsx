'use client';
import React from 'react';
import { useButton } from 'react-aria';
import { ButtonProps } from './Button.interfaces';
import useClassName from '@app/hooks/useClassName';
import { ButtonColor, ButtonVariant } from '@app/theme';

const buttonConfig: Record<ButtonColor, Record<ButtonVariant, string>> = {
  primary: {
    contained:
      'shadow-lg shadow-primary/[.2] text-primary-contrast bg-primary hover:bg-primary/[.66] focus:ring-primary/[0.3] outline-none focus:ring-4',
    text: 'text-primary hover:bg-primary/[.18] focus:bg-primary/[.18] focus:ring-2 outline-none',
    outline:
      'text-primary dark:hover:outline-primary dark:hover:outline-primary focus:bg-primary/[.18] dark:focus:bg-primary/[.4] active:bg-primary/[0.2] focus:ring-primary/[.2] outline outline-2 dark:outline-primary/[.7] outline-primary focus:outline-primary/[.5] focus:ring-4',
  },
  secondary: {
    contained:
      'shadow-lg shadow-secondary/[.2] text-secondary-contrast bg-secondary hover:bg-secondary/[.66] focus:ring-secondary/[0.5] outline-none focus:ring-4',
    text: 'text-secondary hover:bg-secondary/[.18] focus:bg-secondary/[.18] focus:ring-2 outline-none',
    outline:
      'text-secondary dark:hover:outline-secondary dark:hover:outline-secondary focus:bg-secondary/[.18] dark:focus:bg-secondary/[.4] active:bg-secondary/[0.2] focus:ring-secondary/[.2] outline outline-2 dark:outline-secondary/[.7] outline-secondary focus:outline-secondary/[.5] focus:ring-4',
  },
};

const disabledConfig: Record<ButtonVariant, string> = {
  contained: 'shadow-lg shadow-disabled/[.2] text-hint bg-disabled',
  text: 'text-hint',
  outline: 'text-hint outline outline-2 outline-disabled',
};

export default function Button(props: ButtonProps) {
  const {
    children,
    color = 'primary',
    disabled,
    variant = 'text',
    icon,
    iconPlacement = 'left',
  } = props;
  const ref = React.useRef<HTMLButtonElement>(null);
  const { buttonProps } = useButton(
    { ...props, isDisabled: disabled, elementType: 'button' },
    ref,
  );
  const className = useClassName(
    `select-none px-4 py-2 rounded-lg text-variant-button transition ease-out hover:duration-250 duration-250 motion-reduce:transition-none motion-reduce:hover:transition-none flex flex-row items-center justify-center gap-2 ${
      disabled ? disabledConfig[variant] : buttonConfig[color][variant]
    }`,
    props,
  );
  return (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    <button {...(buttonProps as any)} className={className} ref={ref}>
      {icon && iconPlacement === 'left' && <div>{icon}</div>}
      {children}
      {icon && iconPlacement === 'right' && <div>{icon}</div>}
    </button>
  );
}
