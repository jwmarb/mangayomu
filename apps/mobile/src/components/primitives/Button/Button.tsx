import {
  PressableProps as NativeButtonProps,
  PressableAndroidRippleConfig,
  View,
} from 'react-native';
import { styles, variants } from '@/components/primitives/Button/styles';
import {
  BUTTON_COLORS,
  BUTTON_VARIANTS,
  ButtonColors,
  ButtonVariants,
} from '@/components/primitives/types';
import useStyles from '@/hooks/useStyles';
import Pressable from '@/components/primitives/Pressable';
import { createThemedProps } from '@/utils/theme';
import useThemedProps from '@/hooks/useThemedProps';
import Text from '@/components/primitives/Text';
import useContrast from '@/hooks/useContrast';

export type ButtonProps = NativeButtonProps & {
  variant?: ButtonVariants;
  color?: ButtonColors;
  contrast?: boolean;
  textAlignment?: 'left' | 'center' | 'right';
  title: string;
};

const themedProps = (color: ButtonColors) =>
  createThemedProps((theme) => ({
    android_ripple: {
      color: theme.palette[color].ripple,
    } as PressableAndroidRippleConfig,
  }));

const composedVariants = BUTTON_COLORS.reduce((prev, curr) => {
  prev[curr] = variants(curr);
  return prev;
}, {} as Record<ButtonColors, ReturnType<typeof variants>>);

const composedRippleColors = BUTTON_COLORS.reduce((prev, curr) => {
  prev[curr] = themedProps(curr);
  return prev;
}, {} as Record<ButtonColors, ReturnType<typeof themedProps>>);

const composedStyles = BUTTON_VARIANTS.reduce((prev1, curr1) => {
  prev1[curr1] = BUTTON_COLORS.reduce((prev2, curr2) => {
    prev2[curr2] = styles(curr1, curr2);
    return prev2;
  }, {} as Record<ButtonColors, ReturnType<typeof styles>>);
  return prev1;
}, {} as Record<ButtonVariants, Record<ButtonColors, ReturnType<typeof styles>>>);

export default function Button(props: ButtonProps) {
  const {
    variant: variantProp = 'text',
    color = 'primary',
    title,
    contrast: contrastProp,
    textAlignment = 'left',
    ...rest
  } = props;
  const contrast = useContrast(contrastProp);
  const variant = useStyles(composedVariants[color], contrast);
  const btnStyles = useStyles(composedStyles[variantProp][color], contrast);
  const { android_ripple } = useThemedProps(
    composedRippleColors[color],
    contrast,
  );
  const style = [variant[variantProp], btnStyles.button];
  const textColor = variantProp === 'contained' ? 'textPrimary' : color;

  return (
    <View style={btnStyles.container}>
      <Pressable android_ripple={android_ripple} style={style} {...rest}>
        <Text alignment={textAlignment} variant="button" color={textColor}>
          {title}
        </Text>
      </Pressable>
    </View>
  );
}
