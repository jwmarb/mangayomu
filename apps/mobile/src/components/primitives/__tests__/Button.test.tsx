import { render, userEvent } from '@testing-library/react-native';
import Button from '@/components/primitives/Button';

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.useRealTimers();
});

test('renders properly', () => {
  let tree = render(<Button title="My Button Component" />);
  expect(tree).toMatchSnapshot();

  tree = render(<Button title="My Button Component" variant="contained" />);
  expect(tree).toMatchSnapshot();

  tree = render(<Button title="My Button Component" variant="outlined" />);
  expect(tree).toMatchSnapshot();
});

test('interaction works', async () => {
  const onPress = jest.fn();
  const onLongPress = jest.fn();
  const user = userEvent.setup();
  const tree = render(
    <Button
      title="My Button Component"
      onPress={onPress}
      onLongPress={onLongPress}
      testID="btn"
    />,
  );

  const button = tree.getByTestId('btn');
  await user.press(button);
  await user.longPress(button);

  jest.runOnlyPendingTimers();
  expect(onPress).toHaveBeenCalled();
  expect(onLongPress).toHaveBeenCalled();
});
