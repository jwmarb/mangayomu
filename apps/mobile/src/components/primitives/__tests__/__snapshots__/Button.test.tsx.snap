// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<View
  style={
    {
      "borderColor": "transparent",
      "borderRadius": 32,
      "borderWidth": 1.5,
      "overflow": "hidden",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          },
          [
            {},
            {
              "paddingHorizontal": 20,
              "paddingVertical": 12,
            },
          ],
        ],
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 14,
            "fontWeight": "500",
            "letterSpacing": 0.3,
          },
          {
            "color": "#1996FD",
          },
          {
            "textAlign": "left",
          },
          [],
          [],
          undefined,
        ]
      }
    >
      My Button Component
    </Text>
  </View>
</View>
`;

exports[`renders properly 2`] = `
<View
  style={
    {
      "borderColor": "transparent",
      "borderRadius": 32,
      "borderWidth": 1.5,
      "overflow": "hidden",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          },
          [
            {
              "backgroundColor": "#1996FD",
            },
            {
              "paddingHorizontal": 20,
              "paddingVertical": 12,
            },
          ],
        ],
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 14,
            "fontWeight": "500",
            "letterSpacing": 0.3,
          },
          {
            "color": "rgba(0, 0, 0, 0.87)",
          },
          {
            "textAlign": "left",
          },
          [],
          [],
          undefined,
        ]
      }
    >
      My Button Component
    </Text>
  </View>
</View>
`;

exports[`renders properly 3`] = `
<View
  style={
    {
      "borderColor": "#1996FD",
      "borderRadius": 32,
      "borderWidth": 1.5,
      "overflow": "hidden",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          },
          [
            {
              "borderColor": "#1996FD",
            },
            {
              "paddingHorizontal": 20,
              "paddingVertical": 12,
            },
          ],
        ],
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 14,
            "fontWeight": "500",
            "letterSpacing": 0.3,
          },
          {
            "color": "#1996FD",
          },
          {
            "textAlign": "left",
          },
          [],
          [],
          undefined,
        ]
      }
    >
      My Button Component
    </Text>
  </View>
</View>
`;
